{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e0f7dea0",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ea9ce09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# just run the first time notebook is used. Python version at least 3.8 (did not install successfully with 3.7)\n",
    "#pip install malariagen_data==7.1.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc0850c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.8.15\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f0415635",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant packages\n",
    "from pyprojroot import here\n",
    "import pandas as pd\n",
    "import gcsfs\n",
    "import zarr\n",
    "from dask_gateway import Gateway\n",
    "import dask\n",
    "import dask.array as da\n",
    "import numpy as np\n",
    "from fsspec.implementations.zip import ZipFileSystem\n",
    "import os\n",
    "import allel\n",
    "import malariagen_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1c553c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define the chromosomes and crosses to be investigated\n",
    "chroms=[\"2L\",\"2R\",\"3L\",\"3R\",\"X\"]\n",
    "crosses=[\"B1\", \"B3\", \"B5\", \"K2\", \"K4\", \"K6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "db8ed761",
   "metadata": {},
   "outputs": [],
   "source": [
    "#finish off the few that didn't run\n",
    "chroms=[\"2R\"]\n",
    "crosses=[\"K2\", \"K4\", \"K6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d3337800",
   "metadata": {},
   "outputs": [],
   "source": [
    "#and then the 3rd chromosome\n",
    "chroms=[\"3L\",\"3R\"]\n",
    "crosses=[\"B1\", \"B3\", \"B5\", \"K2\", \"K4\", \"K6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2cf9c394",
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrom in chroms:\n",
    "    for cross in crosses:\n",
    "        cross=cross\n",
    "        contig=chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e956655",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.0.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.0.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.0.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Sets up the data\n",
    "ag3 = malariagen_data.Ag3(\"gs://vo_agam_release/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1617725",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sets up paths to gcs buckets\n",
    "gcs_source_zarr_path_template = 'vo_agam_release/v3/snp_genotypes/per_sample/AG1000G-X/zarr/{sample}.zarr.zip'\n",
    "gcs_output_dir_path_template = 'vo_agam_release/v3/snp_genotypes/per_sample/AG1000G-X/zarr/potential_fathers_combined.zarr'\n",
    "gcs_final_output_zarr_path = 'vo_agam_release/v3/snp_genotypes/per_sample/AG1000G-X/zarr/potential_fathers_filtered.zarr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b0369054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sets up access to gcs - project might need to be modified\n",
    "gcs = gcsfs.GCSFileSystem(\n",
    "    project='malariagen-jupyterhub', \n",
    "    token='anon',\n",
    "    block_size=2**18,  # minimum, reduces memory footprint when reading from zip files\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef5998b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sets up access to sample lists - depends on local file system\n",
    "df_crosses = pd.read_csv(here() / \"metadata_original_samples\" / \"original_samples.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a58e29ab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AC0335-C', 'AC0336-C', 'AC0337-C', 'AC0338-C', 'AC0339-C', 'AC0340-C', 'AC0341-C', 'AC0342-C', 'AC0343-C', 'AC0344-C', 'AC0345-C', 'AC0346-C', 'AC0347-C', 'AC0348-C', 'AC0349-C']\n"
     ]
    }
   ],
   "source": [
    "# Defines sample lists for the cross\n",
    "f2_partner_sample_id = [psi[:2] for psi in df_crosses.partner_sample_id]\n",
    "offsprings = [f2_psi == cross for f2_psi in f2_partner_sample_id]\n",
    "samples = list(df_crosses[offsprings].original_sample_id + '-C')\n",
    "print(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3868e832",
   "metadata": {},
   "source": [
    "# Function definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9e683831",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions dealing with opening input zarr files\n",
    "def open_input(samples, seq_id, field):\n",
    "    \n",
    "    if isinstance(samples, list):\n",
    "        zl = []\n",
    "        for s in samples:\n",
    "            zl.append(_open_input(s, seq_id, field))\n",
    "        return da.concatenate(zl, axis=1)\n",
    "    elif isinstance(samples, str):\n",
    "        return _open_input(samples, seq_id, field)\n",
    "    else:\n",
    "        raise TypeError(type(samples))\n",
    "        \n",
    "        \n",
    "def _open_input(sample, seq_id, field):\n",
    "    gcs_path = gcs_source_zarr_path_template.format(sample=sample)\n",
    "    array_path = f'{sample}/{seq_id}/{field}'\n",
    "\n",
    "    # Get the zarr store from the zip file\n",
    "    zip_file = gcs.open(gcs_path)\n",
    "    zip_fs = ZipFileSystem(zip_file) \n",
    "    zarr_store = zarr.storage.KVStore(zip_fs.get_mapper(\"/\"))\n",
    "\n",
    "    # open and load the array\n",
    "    root = zarr.open(store=zarr_store, mode='r')\n",
    "    z = root[array_path][:]\n",
    "    \n",
    "    return z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "731df054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checks that the input us correctly accessed\n",
    "def check_array_setup(samples, seq_id, field):\n",
    "    \n",
    "    # Returns information about data array for the first sample for the specified \n",
    "    # sample set, seq_id and field.\n",
    "\n",
    "    z = open_input(samples=samples[0], seq_id=seq_id, field=field)\n",
    "    return z.dtype, z.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1911193c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility function\n",
    "def my_not(a):\n",
    "    return not a\n",
    "\n",
    "not_arr = np.vectorize(my_not)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d183e1",
   "metadata": {},
   "source": [
    "# Main code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9d94187e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#choose a data set to try out the code\n",
    "cross=\"K2\"\n",
    "contig=\"2L\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "816255af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(dtype('int8'), (48525747, 1, 2))"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checks that the data is correctly accessed\n",
    "\n",
    "check_array_setup(samples=samples, seq_id = contig, field = 'calldata/GT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "99b47c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Accesses the data\n",
    "def access_data(samples=samples, contig=contig, field='calldata/GT'):\n",
    "    z = open_input(samples=samples, seq_id = contig, field = 'calldata/GT')\n",
    "    return z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c51d8853",
   "metadata": {},
   "outputs": [],
   "source": [
    "z=access_data(samples=samples, contig=contig, field='calldata/GT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "cc6bf96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Casts the data into a Genotype Array\n",
    "def make_genotype_array(z=z):\n",
    "    gt = allel.GenotypeDaskArray(z)\n",
    "    return gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a2cd4404",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt=make_genotype_array(z=z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f5485c54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rechunks the data to make computation more efficient\n",
    "def rechunk_data(gt=gt):\n",
    "    gt_r = gt.rechunk((66580, 9, 2))\n",
    "    gt_r_loc = gt_r.compute()\n",
    "    return gt_r, gt_r_loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9f87e5fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_r, gt_r_loc=rechunk_data(gt=gt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a00ed4a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Computes allele counts\n",
    "def compute_allele_counts(gt_r_loc=gt_r_loc):\n",
    "    ac_loc = gt_r_loc.count_alleles()\n",
    "    return ac_loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a65e38db",
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_loc=compute_allele_counts(gt_r_loc=gt_r_loc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b9747ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identifies triallelic sites\n",
    "def identify_triallelic_sites(ac_loc=ac_loc):\n",
    "    is_triallel = (ac_loc.allelism() > 2)\n",
    "    return is_triallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5cc450c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_triallel=identify_triallelic_sites(ac_loc=ac_loc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1abd9460",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sets up the result array and identifies heterozygous, homozygous and missing sites/samples\n",
    "def results_array(gt_r=gt_r, gt_r_loc=gt_r_loc):\n",
    "    res = np.zeros((gt_r_loc.shape[0], 2), dtype='i4')\n",
    "    gt_het = gt_r_loc.is_het()\n",
    "    gt_hom = gt_r_loc.is_hom()\n",
    "    gt_miss = gt_r_loc.is_missing()\n",
    "    return res, gt_het, gt_hom, gt_miss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "32892ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "res, gt_het, gt_hom, gt_miss=results_array(gt_r=gt_r, gt_r_loc=gt_r_loc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "44ce8110",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identifies sites where some samples are homozygous and sites where any samples are missing. Thisnis more stringent than the\n",
    "#old version where it was required that there were fewer than 5 sibs with missing data at each snp\n",
    "def mixed_sites(gt_hom=gt_hom, gt_miss=gt_miss):\n",
    "    is_some_hom = (da.sum(gt_hom, axis = 1).compute() >= 1)\n",
    "    is_miss = (da.sum(gt_miss, axis = 1).compute() >= 1)\n",
    "    return is_some_hom, is_miss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e106998e",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_some_hom, is_miss=mixed_sites(gt_hom=gt_hom, gt_miss=gt_miss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "da4cd294",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Computes actual data\n",
    "def compute_actual_data(gt_r_loc=gt_r_loc, \\\n",
    "                        ac_loc=ac_loc, \\\n",
    "                        is_miss=is_miss, \\\n",
    "                        is_triallel=is_triallel, \\\n",
    "                        is_some_hom=is_some_hom, \\\n",
    "                        gt_miss=gt_miss, \\\n",
    "                        gt_hom=gt_hom, \\\n",
    "                        gt_het=gt_het):\n",
    "    for i in range(gt_r_loc.shape[0]):\n",
    "        if is_miss[i]: # If any missing data missing, return missing\n",
    "            res[i] = np.array([-1,-1])\n",
    "        elif is_triallel[i]: # if triallelic, return homozygous for most frequent allel\n",
    "            maj_val = np.argmax(ac_loc[i])\n",
    "            res[i] = np.array([maj_val,maj_val])\n",
    "        elif is_some_hom[i]: # If some are homozygous, return the homozygous allel\n",
    "            res[i] = gt_r_loc[i][gt_hom[i]][0]\n",
    "        else: #Otherwise, return the heterozygous\n",
    "            not_miss = not_arr(gt_miss[i])\n",
    "            res[i] = gt_r_loc[i][not_miss][0]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "eedefabb",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = compute_actual_data(gt_r_loc=gt_r_loc, \\\n",
    "                        ac_loc=ac_loc, \\\n",
    "                        is_miss=is_miss, \\\n",
    "                        is_triallel=is_triallel, \\\n",
    "                        is_some_hom=is_some_hom, \\\n",
    "                        gt_miss=gt_miss, \\\n",
    "                        gt_hom=gt_hom, \\\n",
    "                        gt_het=gt_het)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a5e01e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Accesses data on sites\n",
    "def access_site_data(contig=contig):\n",
    "    pos = ag3.snp_sites(contig,'POS').compute()\n",
    "    ref = ag3.snp_sites(contig,'REF').compute()\n",
    "    alt = ag3.snp_sites(contig,'ALT').compute()\n",
    "    return pos, ref, alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "78c4dc42",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos, ref, alt =access_site_data(contig=contig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8bf37db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the template (the example header file needs to be in the same directory as this worksheet)\n",
    "def open_template():\n",
    "    template = open(\"AC0300-C_header.txt\", \"r\")\n",
    "\n",
    "    lines = template.read().split('\\n')\n",
    "\n",
    "    template.close()\n",
    "    return lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "60946a80",
   "metadata": {},
   "outputs": [],
   "source": [
    "lines = open_template()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0c662b49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updates the template for this specific run\n",
    "def update_template(lines=lines, contig=contig, cross=cross):\n",
    "        \n",
    "    vcom = lines[-3].split(' ')\n",
    "    head = lines[-2].split('\\t')\n",
    "\n",
    "    vcom[-1] = cross + \"-\" + contig + \".vcf.gz\"\n",
    "    head[-1] = cross \n",
    "\n",
    "    lines[-3] = \" \".join(vcom)\n",
    "    lines[-2] = \"\\t\".join(head)\n",
    "    return lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cc9d4d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lines = update_template(lines=lines, contig=contig, cross=cross)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2a57deed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defines the output file name \n",
    "def output_file_name(prefix, cross=cross, contig=contig):\n",
    "    file_name = prefix + cross + \"-\" + contig + \".vcf\"\n",
    "    return file_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8943c79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name=output_file_name(prefix=\"stringent\", cross=cross, contig=contig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f891752a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'stringentK2-2L.vcf'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "686a2637",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writes header and data to file\n",
    "def write_vcf(file_name=file_name, lines=lines, pos=pos, contig=contig, res=res, ref=ref, alt=alt):\n",
    "        res_file = open(file_name, \"w\")\n",
    "        res_file.write(\"\\n\".join([line for line in lines[:-1]]))\n",
    "        for i in range(0,len(pos)):#100):\n",
    "            res_file.write(\"\\n\" + contig + \"\\t\" + str(pos[i]) + \"\\t.\\t\" + ref[i].decode(\"utf-8\") \\\n",
    "                           + \"\\t\" + alt[i,0].decode(\"utf-8\") + \",\" + alt[i,1].decode(\"utf-8\") + \",\" \\\n",
    "                           + alt[i,2].decode(\"utf-8\") + \"\\t.\\t.\\t.\\tGT:AD:DP:GQ:PL:SB\\t\" + str(res[i,0]) \\\n",
    "                           + \"/\" + str(res[i,1]) + \":.:.:.:.:.\")\n",
    "        res_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "108fd7fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_vcf(file_name=file_name, lines=lines, pos=pos, contig=contig, res=res, ref=ref, alt=alt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "15a5b808",
   "metadata": {},
   "outputs": [],
   "source": [
    "def file_check(file_name=file_name):\n",
    "    file_check = open(file_name, \"r\")\n",
    "\n",
    "    file_check_lines = file_check.read(10_000).split('\\n')\n",
    "\n",
    "    file_check.close()\n",
    "    return file_check_lines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1c040de",
   "metadata": {},
   "source": [
    "# Put the functions together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29435632",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loop through the chromosomes and crosses to be investigated, writing a vcf file for each. VCF files written in same\n",
    "#directory as this worksheet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e4212d1b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B1\n",
      "3L\n",
      "['AC0352-C', 'AC0353-C', 'AC0354-C', 'AC0355-C', 'AC0356-C', 'AC0357-C', 'AC0358-C', 'AC0359-C', 'AC0360-C']\n",
      "data accessed forB1and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]]\n",
      "[[False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B1-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB1-3L', '']\n",
      "file name iscrossB1-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "B3\n",
      "3L\n",
      "['AC0362-C', 'AC0363-C', 'AC0364-C', 'AC0365-C', 'AC0366-C', 'AC0367-C', 'AC0368-C', 'AC0369-C', 'AC0370-C', 'AC0371-C', 'AC0372-C', 'AC0373-C', 'AC0374-C', 'AC0375-C', 'AC0376-C', 'AC0377-C', 'AC0378-C', 'AC0379-C', 'AC0380-C', 'AC0381-C']\n",
      "data accessed forB3and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B3-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB3-3L', '']\n",
      "file name iscrossB3-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "B5\n",
      "3L\n",
      "['AC0383-C', 'AC0384-C', 'AC0385-C', 'AC0386-C', 'AC0387-C', 'AC0388-C', 'AC0389-C', 'AC0390-C', 'AC0391-C', 'AC0392-C', 'AC0393-C']\n",
      "data accessed forB5and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]]\n",
      "[[False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B5-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB5-3L', '']\n",
      "file name iscrossB5-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K2\n",
      "3L\n",
      "['AC0301-C', 'AC0302-C', 'AC0303-C', 'AC0304-C', 'AC0305-C', 'AC0306-C', 'AC0307-C', 'AC0308-C', 'AC0309-C', 'AC0310-C', 'AC0311-C', 'AC0312-C', 'AC0313-C', 'AC0314-C', 'AC0315-C', 'AC0316-C']\n",
      "data accessed forK2and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K2-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK2-3L', '']\n",
      "file name iscrossK2-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K4\n",
      "3L\n",
      "['AC0318-C', 'AC0319-C', 'AC0320-C', 'AC0321-C', 'AC0322-C', 'AC0323-C', 'AC0324-C', 'AC0325-C', 'AC0326-C', 'AC0327-C', 'AC0328-C', 'AC0329-C', 'AC0330-C', 'AC0331-C', 'AC0332-C', 'AC0333-C']\n",
      "data accessed forK4and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K4-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK4-3L', '']\n",
      "file name iscrossK4-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K6\n",
      "3L\n",
      "['AC0335-C', 'AC0336-C', 'AC0337-C', 'AC0338-C', 'AC0339-C', 'AC0340-C', 'AC0341-C', 'AC0342-C', 'AC0343-C', 'AC0344-C', 'AC0345-C', 'AC0346-C', 'AC0347-C', 'AC0348-C', 'AC0349-C']\n",
      "data accessed forK6and3L\n",
      "genotype_array_made\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "data_rechunked\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "./. ./. ./. ./. ./. ... ./. ./. ./. ./. ./.\n",
      "\n",
      "allele counts computed\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "0 0 0 0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]]\n",
      "calculated mixed sites\n",
      "[False False False False False]\n",
      "[ True  True  True  True  True]\n",
      "actual data computed\n",
      "[[-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]\n",
      " [-1 -1]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'A' b'C' b'G' b'A' b'A']\n",
      "[[b'C' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'T']\n",
      " [b'C' b'T' b'G']\n",
      " [b'C' b'T' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K6-3L.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK6-3L', '']\n",
      "file name iscrossK6-3L.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "B1\n",
      "3R\n",
      "['AC0352-C', 'AC0353-C', 'AC0354-C', 'AC0355-C', 'AC0356-C', 'AC0357-C', 'AC0358-C', 'AC0359-C', 'AC0360-C']\n",
      "data accessed forB1and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "18  0  0  0\n",
      "18  0  0  0\n",
      "18  0  0  0\n",
      "18  0  0  0\n",
      "18  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True]]\n",
      "[[False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]\n",
      " [False False False False False False False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B1-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB1-3R', '']\n",
      "file name iscrossB1-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "B3\n",
      "3R\n",
      "['AC0362-C', 'AC0363-C', 'AC0364-C', 'AC0365-C', 'AC0366-C', 'AC0367-C', 'AC0368-C', 'AC0369-C', 'AC0370-C', 'AC0371-C', 'AC0372-C', 'AC0373-C', 'AC0374-C', 'AC0375-C', 'AC0376-C', 'AC0377-C', 'AC0378-C', 'AC0379-C', 'AC0380-C', 'AC0381-C']\n",
      "data accessed forB3and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "40  0  0  0\n",
      "40  0  0  0\n",
      "40  0  0  0\n",
      "40  0  0  0\n",
      "40  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True  True  True  True  True]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False False False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B3-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB3-3R', '']\n",
      "file name iscrossB3-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "B5\n",
      "3R\n",
      "['AC0383-C', 'AC0384-C', 'AC0385-C', 'AC0386-C', 'AC0387-C', 'AC0388-C', 'AC0389-C', 'AC0390-C', 'AC0391-C', 'AC0392-C', 'AC0393-C']\n",
      "data accessed forB5and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "22  0  0  0\n",
      "22  0  0  0\n",
      "22  0  0  0\n",
      "22  0  0  0\n",
      "22  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True]]\n",
      "[[False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]\n",
      " [False False False False False False False False False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only B5-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tB5-3R', '']\n",
      "file name iscrossB5-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K2\n",
      "3R\n",
      "['AC0301-C', 'AC0302-C', 'AC0303-C', 'AC0304-C', 'AC0305-C', 'AC0306-C', 'AC0307-C', 'AC0308-C', 'AC0309-C', 'AC0310-C', 'AC0311-C', 'AC0312-C', 'AC0313-C', 'AC0314-C', 'AC0315-C', 'AC0316-C']\n",
      "data accessed forK2and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K2-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK2-3R', '']\n",
      "file name iscrossK2-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K4\n",
      "3R\n",
      "['AC0318-C', 'AC0319-C', 'AC0320-C', 'AC0321-C', 'AC0322-C', 'AC0323-C', 'AC0324-C', 'AC0325-C', 'AC0326-C', 'AC0327-C', 'AC0328-C', 'AC0329-C', 'AC0330-C', 'AC0331-C', 'AC0332-C', 'AC0333-C']\n",
      "data accessed forK4and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "32  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True  True]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K4-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK4-3R', '']\n",
      "file name iscrossK4-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n",
      "K6\n",
      "3R\n",
      "['AC0335-C', 'AC0336-C', 'AC0337-C', 'AC0338-C', 'AC0339-C', 'AC0340-C', 'AC0341-C', 'AC0342-C', 'AC0343-C', 'AC0344-C', 'AC0345-C', 'AC0346-C', 'AC0347-C', 'AC0348-C', 'AC0349-C']\n",
      "data accessed forK6and3R\n",
      "genotype_array_made\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "data_rechunked\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
      "\n",
      "allele counts computed\n",
      "30  0  0  0\n",
      "30  0  0  0\n",
      "30  0  0  0\n",
      "30  0  0  0\n",
      "30  0  0  0\n",
      "\n",
      "triallelic_sites_identified\n",
      "[False False False False False]\n",
      "made results array\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]]\n",
      "[[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]\n",
      " [ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "   True  True  True]]\n",
      "[[False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]\n",
      " [False False False False False False False False False False False False\n",
      "  False False False]]\n",
      "calculated mixed sites\n",
      "[ True  True  True  True  True]\n",
      "[False False False False False]\n",
      "actual data computed\n",
      "[[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "site data accessed\n",
      "[1 2 3 4 5]\n",
      "[b'C' b'C' b'T' b'C' b'T']\n",
      "[[b'A' b'T' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']\n",
      " [b'A' b'T' b'G']\n",
      " [b'A' b'C' b'G']]\n",
      "template updated\n",
      "['##fileformat=VCFv4.2', '##FILTER=<ID=PASS,Description=\"All filters passed\">', '##FILTER=<ID=LowQual,Description=\"Low quality\">', '##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">', '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">', '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">', '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">', '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">', '##GATKCommandLine.UnifiedGenotyper=<ID=UnifiedGenotyper,Version=3.7-0-gcfedb67,Date=\"Fri Mar 08 08:41:32 GMT 2019\",Epoch=1552034492817,CommandLineOptions=\"analysis_type=UnifiedGenotyper input_file=[/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/bams.list] showFullBamList=false read_buffer_size=null read_filter=[] disable_read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 secondsBetweenProgressUpdates=10 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=false bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=DYNAMIC_SEEK variant_index_parameter=-1 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=BOTH pcr_error_rate=1.0E-4 computeSLOD=false pair_hmm_implementation=LOGLESS_CACHING min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 annotateNDA=false useNewAFCalculator=false heterozygosity=0.015 indel_heterozygosity=0.001 heterozygosity_stdev=0.05 standard_min_confidence_threshold_for_calling=0.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=6 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=GENOTYPE_GIVEN_ALLELES alleles=(RodBinding name=alleles source=/lustre/scratch118/malaria/team112/pipelines/resources/vobs/callsets/ag.allsites.nonN.vcf.gz) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_ALL_SITES allSitePLs=false dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=/lustre/scratch118/malaria/team112/pipelines/setups/vo_agam/output/7/4/a/a/1248689/4_gatk_unified_genotyper_gatk3_v3/gatk.vcf.gz onlyEmitSamples=[] debug_file=null metrics_file=null annotation=[RMSMappingQuality, DepthPerAlleleBySample] excludeAnnotation=[Coverage, MappingQualityZero, HaplotypeScore, SpanningDeletions, FisherStrand, StrandOddsRatio, ChromosomeCounts, BaseQualityRankSumTest, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest] filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\">', '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">', '##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">', '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">', '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">', '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">', '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">', '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">', '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">', '##contig=<ID=2R,length=61545105>', '##contig=<ID=3R,length=53200684>', '##contig=<ID=2L,length=49364325>', '##contig=<ID=UNKN,length=42389979>', '##contig=<ID=3L,length=41963435>', '##contig=<ID=X,length=24393108>', '##contig=<ID=Y_unplaced,length=237045>', '##contig=<ID=Mt,length=15363>', '##reference=file:///lustre/scratch118/malaria/team112/pipelines/resources/vobs/anopheles-gambiae-pestchromosomesagamp4.fa', '##bcftools_viewVersion=1.3.1+htslib-1.3.1', '##bcftools_viewCommand=view --header-only K6-3R.vcf.gz', '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tK6-3R', '']\n",
      "file name iscrossK6-3R.vcf\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf file written\n",
      "file closed\n"
     ]
    }
   ],
   "source": [
    "for chrom in chroms:\n",
    "    for cross in crosses:\n",
    "        cross=cross\n",
    "        print(cross)\n",
    "        contig=chrom\n",
    "        seqid=contig\n",
    "        print(chrom)\n",
    "        f2_partner_sample_id = [psi[:2] for psi in df_crosses.partner_sample_id]\n",
    "        offsprings = [f2_psi == cross for f2_psi in f2_partner_sample_id]\n",
    "        samples = list(df_crosses[offsprings].original_sample_id + '-C')\n",
    "        print(samples)\n",
    "        z= access_data(samples=samples, contig=contig, field='calldata/GT')\n",
    "        print(\"data accessed for\"+cross+\"and\"+chrom)\n",
    "        gt= make_genotype_array(z=z)\n",
    "        print(\"genotype_array_made\")\n",
    "        print(gt[:5])\n",
    "        gt_r, gt_r_loc=rechunk_data(gt=gt)\n",
    "        print(\"data_rechunked\")\n",
    "        print(gt_r[:5])\n",
    "        print(gt_r_loc[:5])\n",
    "        ac_loc=compute_allele_counts(gt_r_loc=gt_r_loc)\n",
    "        print(\"allele counts computed\")\n",
    "        print(ac_loc[:5])\n",
    "        is_triallel=identify_triallelic_sites(ac_loc=ac_loc)\n",
    "        print(\"triallelic_sites_identified\")\n",
    "        print(is_triallel[:5])\n",
    "        res, gt_het, gt_hom, gt_miss = results_array(gt_r=gt_r, gt_r_loc=gt_r_loc)\n",
    "        print(\"made results array\")\n",
    "        print(res[:5])\n",
    "        print(gt_het[:5])\n",
    "        print(gt_hom[:5])\n",
    "        print(gt_miss[:5])\n",
    "        is_some_hom, is_miss = mixed_sites(gt_hom=gt_hom, gt_miss=gt_miss)\n",
    "        print(\"calculated mixed sites\")\n",
    "        print(is_some_hom[:5])\n",
    "        print(is_miss[:5])\n",
    "        res = compute_actual_data(gt_r_loc=gt_r_loc, \\\n",
    "                        ac_loc=ac_loc, \\\n",
    "                        is_miss=is_miss, \\\n",
    "                        is_triallel=is_triallel, \\\n",
    "                        is_some_hom=is_some_hom, \\\n",
    "                        gt_miss=gt_miss, \\\n",
    "                        gt_hom=gt_hom, \\\n",
    "                        gt_het=gt_het)\n",
    "        print(\"actual data computed\")\n",
    "        print(res[:5])\n",
    "        pos, ref, alt = access_site_data(contig=contig)\n",
    "        print(\"site data accessed\")\n",
    "        print(pos[:5])\n",
    "        print(ref[:5])\n",
    "        print(alt[:5])\n",
    "        lines = open_template()\n",
    "        lines = update_template(lines=lines, contig=contig, cross=cross)\n",
    "        print(\"template updated\")\n",
    "        print(lines)\n",
    "        file_name = output_file_name(prefix=\"cross\", cross=cross, contig=contig)\n",
    "        print(\"file name is\" + file_name)\n",
    "        write_vcf(file_name=file_name, lines=lines, pos=pos, contig=contig, res=res, ref=ref, alt=alt)\n",
    "        print(\"vcf file written\")\n",
    "        print(\"file closed\")\n",
    "        file_check()\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cd8a166",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use bcftools concat to concatenate chromosomes\n",
    "#bcftools concat -o file_concat.vcf.gz -Oz file_2R.vcf file_3R.vcf file_2L.vcf file_3L.vcf\n",
    "#ignore X for now because with male sibs can't reconstruct genotype\n",
    "#tabix -p vcf file.vcf.gz\n",
    "#continue to use the vcf files with ASERead_counter_star as \"vcfjoind\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
